apis/buzzer/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+//! Implementation started by : https://github.com/teodorobert
+//! Continued and modified by : https://github.com/SheepSeb
+#![no_std]
+
+use core::cell::Cell;
+use core::time::Duration;
+
+use libtock_platform::{share, DefaultConfig, ErrorCode, Subscribe, Syscalls};
+pub struct Buzzer<S: Syscalls>(S);
+
+impl<S: Syscalls> Buzzer<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// Initiate a tone
+    pub fn tone(freq: u32, duration: Duration) -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, BUZZER_ON, freq, duration.as_millis() as u32).to_result()
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share>(
+        listener: &'share Cell<Option<(u32,)>>,
+        subscribe: share::Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Initiate a synchronous tone
+    /// Returns Ok() if the operation was successful
+    pub fn tone_sync(freq: u32, duration: Duration) -> Result<u32, ErrorCode> {
+        let listener = Cell::new(Some((0,)));
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::tone(freq, duration) {
+                    while listener.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
```

Comment:
Use ?

Reply:
This code is outdated in my branch is like [this](https://github.com/SheepSeb/libtock-rs/blob/sebnae/buzzer/apis/buzzer/src/lib.rs)